database-commons/src/main/java/io/cdap/plugin/db/sink/AbstractDBSink.java

```diff
@@ -302,33 +310,23 @@ private Schema inferSchema(Class<? extends Driver> driverClass) {
                                           dbSinkConfig.getJdbcPluginName());
       Properties connectionProperties = new Properties();
       connectionProperties.putAll(dbSinkConfig.getConnectionArguments());
-      try (Connection connection = DriverManager.getConnection(dbSinkConfig.getConnectionString(),
-                                                               connectionProperties)) {
+      try (Connection connection = RetryUtils.createConnectionWithRetry((RetryPolicy<Connection>) retryPolicy,
+        dbSinkConfig.getConnectionString(), connectionProperties, getErrorDetailsProvider())) {
         executeInitQueries(connection, dbSinkConfig.getInitQueries());
-
-        try (Statement statement = connection.createStatement();
-             ResultSet rs = statement.executeQuery("SELECT * FROM " + fullyQualifiedTableName
-                                                     + " WHERE 1 = 0")) {
+        try (Statement statement = RetryUtils.createStatementWithRetry((RetryPolicy<Statement>) retryPolicy,
+          connection, getErrorDetailsProvider());
+             ResultSet rs = RetryUtils.executeQueryWithRetry((RetryPolicy<ResultSet>) retryPolicy, statement,
+               String.format("SELECT * FROM %s WHERE 1 = 0", fullyQualifiedTableName), getErrorDetailsProvider())) {
           inferredFields.addAll(getSchemaReader().getSchemaFields(rs));
         }
       } catch (SQLException e) {
-        // wrap exception to ensure SQLException-child instances not exposed to contexts w/o jdbc driver in classpath
-        String errorMessage =
-          String.format("SQL Exception occurred: [Message='%s', SQLState='%s', ErrorCode='%s'].", e.getMessage(),
-            e.getSQLState(), e.getErrorCode());
-        String errorMessageWithDetails = String.format("Error while reading table metadata." +
-          "Error message: '%s'. Error code: '%s'. SQLState: '%s'", e.getMessage(), e.getErrorCode(), e.getSQLState());
-        String externalDocumentationLink = getExternalDocumentationLink();
-        if (!Strings.isNullOrEmpty(externalDocumentationLink)) {
-          if (!errorMessage.endsWith(".")) {
-            errorMessage = errorMessage + ".";
-          }
-          errorMessage = String.format("%s For more details, see %s", errorMessageWithDetails, errorMessage);
-        }
-        throw ErrorUtils.getProgramFailureException(new ErrorCategory(ErrorCategory.ErrorCategoryEnum.PLUGIN),
-          errorMessage, errorMessageWithDetails, ErrorType.USER, false, ErrorCodeType.SQLSTATE,
-          e.getSQLState(), externalDocumentationLink, new SQLException(e.getMessage(),
-            e.getSQLState(), e.getErrorCode()));
+        LOG.error("Exception while trying to infer schema of database table {} for connection {}. [SQLState={}, " +
+          "ErrorCode={}]", fullyQualifiedTableName, dbSinkConfig.getConnectionString(), e.getSQLState(),
+            e.getErrorCode(), e);
```

Comment:
Logging and throwing is an anti-pattern. Please don't log if we are already adding it in `failureCollector`.